Sentry integration

sondrelg opened the pull request:
I still need to:

- Bump version
- Write docs

but the core logic for this PR is done, and I think it would be good to review that before writing docs, to make sure you agree with the some of the design choices.

--------------

This PR introduces a new package setting, `INTEGRATIONS` which will be a list of pre-made integrations from the new `integrations.py`. 

The idea for this PR originally, was only to add some functionality for integrating the correlation ID with Sentry, but I think it might be a good idea to design this feature in a way that opens up for more integrations in the future.

The second addition is the Sentry logic, which I've packaged in a class so that we can bundle validation logic and execution logic for each integration, and separate it from other code.

Also made some adjustments to doc-strings and inline comments here and there, but feel free to change them back

JonasKs commented on django_guid/signals.py:
```diff
@@ -14,11 +14,12 @@ def delete_guid(sender: Optional[dict], **kwargs: dict) -> None:
 
-    When a request is finished, delete a requests _guid reference to prevent memory leaks.
+    When a request is finished the delete_guid method is called to remove the request _guid reference.
+    This is important to maintain, to prevent a memory leak.
```
Not really a fan of this change, feels like it's more words to say the same thing?

JonasKs commented on django_guid/middleware.py:
```diff
@@ -29,15 +30,18 @@ def __init__(self, get_response: Callable) -> None:
-        # Without signals there may be a memory leak
+        # `django_guid` must be in installed apps for signals to work and if the signals dont work,
+        # it creates a memory leak, since _guid is never cleaned and can theoretically keep growing infinitely.
+        #
```
Can remove these three lines honestly.

JonasKs commented on django_guid/middleware.py:
```diff
@@ -29,15 +30,18 @@ def __init__(self, get_response: Callable) -> None:
         """
         Fetches the current thread ID from the pool and stores the GUID in the _guid class variable,
         with the thread ID as the key.
+
```
Not really a fan of these added newlines in the docstrings.

JonasKs commented on django_guid/middleware.py:
```diff
@@ -57,10 +65,11 @@ def __call__(self, request: HttpRequest) -> Union[HttpRequest, HttpResponse]:
-        If no value has been set for the current thread yet, we return a default value.
-        :default: Optional value to return if no GUID has been set on the current thread.
+        Fetches the GUID of the current thread from _guid.
 
```
Remove new line imo

JonasKs commented on django_guid/middleware.py:
```diff
@@ -133,21 +141,22 @@ def _get_correlation_id_from_header(self, request: HttpRequest) -> str:
     def _get_id_from_header(self, request: HttpRequest) -> str:
         """
         Checks if the request contains the header specified in the Django settings.
+
```
Remove new line

JonasKs commented on django_guid/middleware.py:
```diff
@@ -14,10 +14,11 @@
+    Checks for a GUID (correlation ID) in an incoming request's headers and assigns it to the thread if found.
+    If none is found, one is generated.
+
+    As every request spawns a new thread, this makes it possible to add logger IDs that only span a single request.
```
Not how it works, and not really needed to be explained in detail here. Each WSGI implementation works differently, but they all have a set of threads available and only use those.

JonasKs commented on django_guid/middleware.py:
```diff
@@ -70,15 +79,15 @@ def set_guid(cls, guid: str) -> None:
         Assigns a GUID to the thread.
 
-        :param guid: str
+        :param guid: The GUID being assigned
```
Nice 👍

JonasKs commented on django_guid/integrations.py:
```diff
@@ -0,0 +1,39 @@
+            )
+
+    @staticmethod
+    def run(self):
```
```suggestion
    def run():
```

JonasKs commented on django_guid/integrations.py:
```diff
@@ -0,0 +1,39 @@
+class SentryIntegration:
+    """
+    This integrations, when added to the django_guid settings,
+    ensures that each requests correlation ID is passed on to Sentry exception logs.
```
Docstring probably needs to be revisited 🍡

sondrelg replied:
done

sondrelg commented:
Added a base class + new tests. Put the base class in the __init__ and moved the sentry integration to `sentry.py`. Everything still works the same.

-------

One thing to note is that the SentryIntegration `run` method currently takes `middleware_context` as an argument - this is the `self` instance of the middleware class. 

This was done intentionally so that we can run all integrations like this in the middleware:

```python
        # Run all integrations
        for integration in settings.INTEGRATIONS:
            logger.debug('Running integration: `%s`', integration.identifier)
            integration.run(self)
```

This might not be the best implementation, but the idea here is that, while the sentry `run` method could have easily looked like this

```python
    def run(self, guid) -> None:
        import sentry_sdk
        with sentry_sdk.configure_scope() as scope:
            logger.debug('Setting Sentry transaction_id to %s', guid)
            scope.set_tag('transaction_id', guid)
```

...instead of this

```python
    def run(self, middleware_context) -> None:
        import sentry_sdk
        with sentry_sdk.configure_scope() as scope:
            guid = middleware_context.get_guid()
            logger.debug('Setting Sentry transaction_id to %s', guid)
            scope.set_tag('transaction_id', guid)
```

... this choice gives custom/new integrations access to the middleware itself, which means it's more flexible and less prone to rewrites. The one thing I think might be "wrong" with this implementation is that `self` probably shouldnt be passed as an arg, but should be inherited. I could explain this to you directly if you prefer.

JonasKs commented on django_guid/integrations/__init__.py:
```diff
@@ -0,0 +1,24 @@
+from django.core.exceptions import ImproperlyConfigured
```
We discussed this between us, but I'll write it down.

I'd like this code to live in `integrations.py` or `base.py`, and rather imported in `__init__.py`.

sondrelg replied:
Moved to base.py

JonasKs commented on django_guid/integrations/__init__.py:
```diff
@@ -0,0 +1,24 @@
+
+    identifier = None  # Name the identifier
+
+    def __init__(self):
```
We have two choices here: 

```suggestion
    def __init__(self, *args, **kwargs):
```

or remove `__init__(self):` and do that validation in the Middleware itself.

I kinda think Sentrys way of doing this is sexy - they have `setup_once()`, we could have the same - but this would be called in `__init__` in the middleware.

Something like this:

```py
    def __init__(self, get_response: Callable) -> None:
        """
        One-time configuration and initialization.
        """
        self.get_response = get_response

        # This logic cannot be moved to config.py because apps are not yet initialized when that is executed
        if not apps.is_installed('django_guid'):
            raise ImproperlyConfigured('django_guid must be in installed apps')

        for integration in settings.INTEGRATIONS:
            logger.debug('Installing integration: `%s`', integration.identifier)
            integration.setup_once(self)
```

or something along those lines

sondrelg replied:
Yep lets do that

sondrelg replied:
Implemented with a `setup` and a `validate` method, where the setup is run during the middleware start-up like you suggested, and the validate method is run during settings initialization

JonasKs commented:
Perfect, just what I had in mind!!

I’m on my phone right now reading through, but passing `self` as an argument seems a bit whacky. Not sure if it really is, though. 

How ever, I think the solution could be just as slick with Django Signals approach. We send the GUID as an argument, but every integration must allow `kwargs`, so that if we have to pass arguments to an integration at a later point we can do so in the future without having to break anything. 

https://docs.djangoproject.com/en/3.0/topics/signals/#receiver-functions

sondrelg commented:
Not sure if this fully addresses the main problem which is that by sending the GUID in middleware.py, we will send a GUID, and only a GUID to all custom integrations

JonasKs commented:
I can’t really see where this is an issue though. The middleware does close to nothing but stores a GUID. If that changes in the future, `kwargs` will allow for changes without breaking backwards compatibility. The internal API can be used if one wants to delete the GUID. (Requires an important, obviously, but it feels less whacky than passing `self`)

I’ll read up on passing `self` as an argument before totally dismissing the idea, but I suspect I’ll find that it’s a “nono”.

JonasKs commented:
Actually, quickly looked into `asyncio`, seems like they’re doing it. Maybe just be me, and it’s totally okay to do it.🤷‍♂️  
I’ll look more into it tomorrow.

JonasKs commented:
I know we talked about this this morning but I like having everything written down:   
Essentially it seems like it’s okay, but I feel like it’s safer to do kwargs considering backwards compatibility and to make it more clear what’s happening. So to send the `guid` and require all functions to allow `kwargs` seems to be the best idea to me.

JonasKs commented on tests/functional/test_integration_base_class.py:
```diff
@@ -0,0 +1,94 @@
+        Settings()
+
+
+def test_base():
```
This is a function calling a class, not a test asserting anything.

sondrelg replied:
it will still fail if you manage to break the base class

sondrelg commented:
I added kwarg and callability-validation, and *put it in the config.py*. Last we spoke, it was in the middleware, but I don't think it belongs there when thinking it through properly.

JonasKs commented on tests/functional/test_integration_base_class.py:
```diff
@@ -0,0 +1,94 @@
+    from django.conf import settings
+    from django_guid.config import Settings
+
+    x = SentryIntegration()
```
Should be clear what it is, `mock_integration` for instance

JonasKs commented on tests/functional/test_integration_base_class.py:
```diff
@@ -0,0 +1,94 @@
+
+    x = SentryIntegration()
+
+    # Non-callable run method
```
Either subtlest these (even though the PyPi package seems a bit immature maybe?) or split into three tests. If you do, make the mock a fixture.  
I can help if you haven’t done this before.

JonasKs commented on tests/functional/test_integration_base_class.py:
```diff
@@ -0,0 +1,94 @@
+    with pytest.raises(ImproperlyConfigured, match='Integration method `run` needs to be made callable for `SentryIntegration`.'):
+        Settings()
+
+    # # Non-callable validate method
```
`# #`

JonasKs commented on tests/functional/test_sentry_integration.py:
```diff
@@ -0,0 +1,46 @@
+
+def test_sentry_integration(client, monkeypatch, caplog):
+    """
+    Tests that the package handles multiple header values by defaulting to one and logging a warning.
```
Not sure if I understand what you mean by this?

JonasKs commented on tests/functional/test_sentry_integration.py:
```diff
@@ -0,0 +1,46 @@
+        ('97c304252fd14b25b72d6aee31565842', 'Received signal `request_finished`'),
+        ('97c304252fd14b25b72d6aee31565842', 'Deleting 97c304252fd14b25b72d6aee31565842 from _guid'),
+    ]
+    assert [(x.correlation_id, x.message) for x in caplog.records] == expected
```
Should probably patch out sentry and assert called with.

JonasKs replied:
fixed
